### PR TITLE
Fix parsing of env variables when properties have underscores

### PR DESCRIPTION
**Pull request by georgew5656**
Fixes [#15875.](https://github.com/apache/druid/issues/15875)

### Description
When running the default docker image, we rely on the logic that transformers env variables into runtime properties for easier configuration (https://github.com/apache/druid/blob/master/distribution/docker/druid.sh#L147)

This will transform a env variable like druid_metadata_storage_type=mysql into a runtime prop entry of druid.metadata.storage.type=mysql.

There is a bug with properties that use underscores like druid.indexer.runner.k8s.podTemplate.index_kafka since those have underscores, they will not get properly replaced. e.g. druid_indexer_runner_k8s_podTemplate_index_kafka will get replaced with druid.indexer.runner.k8s.podTemplate.index.kafka instead of druid.indexer.runner.k8s.podTemplate.index_kafka 

#### Added a forbidden-apis entry ...

Although its not ideal to have underscores in runtime props (convention seems to be that runtime props should be camelCase), there is nothing preventing underscores in runtime props and in this case the runtime props are just matching our task type names.

I fixed this bug by adding the ability to escape underscores in runtime props by including double underscores.
e.g. druid_indexer_runner_k8s_podTemplate_index__kafka will properly be evaluated to druid.indexer.runner.k8s.podTemplate.index_kafka now.

I used a double underscore as the escape character instead of \_ because these character go in the env variable name

I added two additional sed lines to the env variable parser to parse these double underscores.
#### Release note
Allow specifying runtime props with underscores through env variables in the docker image with a escape character.

##### Key changed/added classes in this PR
 * `peon.sh`
 * `druid.sh`

This PR has:

- [X] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [ ] a release note entry in the PR description.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [ ] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [X] been tested in a test Druid cluster.

### CODE REVIEW

**Comment by jreyeshdez** [comment]
@georgew5656 thanks for looking into this. I have been testing your fix however not sure if it works:

```
jreyes@hdez ~ % echo druid_indexer_runner_k8s_podTemplate_index_kinesis | sed -e 's?^\([^=]*\)=.*?\1?g' -e 's?__?$UNDERSCORE?g' -e 's?_?.?g' -e 's?$UNDERSCORE?_?g'
druid.indexer.runner.k8s.podTemplate.index.kinesis
```

**Comment by georgew5656** [comment]
> @georgew5656 thanks for looking into this. I have been testing your fix however not sure if it works:
> 
> ```
> jreyes@hdez ~ % echo druid_indexer_runner_k8s_podTemplate_index_kinesis | sed -e 's?^\([^=]*\)=.*?\1?g' -e 's?__?$UNDERSCORE?g' -e 's?_?.?g' -e 's?$UNDERSCORE?_?g'
> druid.indexer.runner.k8s.podTemplate.index.kinesis
> ```

you need to escape the cases where you actually want the underscore to be a part of the property with a double underscore, 

e.g. druid_indexer_runner_k8s_podTemplate_index__kinesis

**Comment by jreyeshdez** [comment]
> > @georgew5656 thanks for looking into this. I have been testing your fix however not sure if it works:
> > ```
> > jreyes@hdez ~ % echo druid_indexer_runner_k8s_podTemplate_index_kinesis | sed -e 's?^\([^=]*\)=.*?\1?g' -e 's?__?$UNDERSCORE?g' -e 's?_?.?g' -e 's?$UNDERSCORE?_?g'
> > druid.indexer.runner.k8s.podTemplate.index.kinesis
> > ```
> 
> you need to escape the cases where you actually want the underscore to be a part of the property with a double underscore,
> 
> e.g. druid_indexer_runner_k8s_podTemplate_index__kinesis

Oh I see, now makes sense and works. Probably something we would need to document somewhere too.

File: distribution/docker/druid.sh [hunk 1]
```diff
@@ -148,7 +148,7 @@ env | grep ^druid_ | while read evar;
 do
     # Can't use IFS='=' to parse since var might have = in it (e.g. password)
     val=$(echo "$evar" | sed -e 's?[^=]*=??')
-    var=$(echo "$evar" | sed -e 's?^\([^=]*\)=.*?\1?g' -e 's?_?.?g')
+    var=$(echo "$evar" | sed -e 's?^\([^=]*\)=.*?\1?g' -e 's?__?$UNDERSCORE?g' -e 's?_?.?g' -e 's?$UNDERSCORE?_?g')
```
Comment (suneet-s) [inline distribution/docker/druid.sh:151]: Is $UNDERSCORE supposed to be defined somewhere?

**Review by suneet-s — COMMENTED** [review]
+1 about adding a note somewhere in the docs. Since this is most likely relevant in the context of mm-less ingestion, I think documenting this behavior of peon.sh  somewhere in this page would be good - https://github.com/apache/druid/blob/master/docs/development/extensions-contrib/k8s-jobs.md#custom-template-pod-adapter

File: distribution/docker/druid.sh (hunk 1 above)
Comment (georgew5656) [reply distribution/docker/druid.sh:151]: its a placeholder to handle the escape char

**Review by techdocsmith — COMMENTED** [review]
docs LGTM

**Comment by georgew5656** [comment]
gonna merge since this failure seems unrelated

**Pull request merged**